### PR TITLE
Unittests for allocator_memory_strategy.hpp part 2

### PR DESCRIPTION
**Pull request by brawner**
This adds the remaining tests for allocator_memory_strategy.hpp and is directly dependent on #1197. The new commit is https://github.com/ros2/rclcpp/commit/5a88b09bfe0ad7f068f5d53b7f6de04f69f58a65

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11300/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6546/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9258/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11204/)

### CODE REVIEW

**Comment by clalancette** [comment]
> This adds the remaining tests for allocator_memory_strategy.hpp and is directly dependent on #1197. The new commit is [5a88b09](https://github.com/ros2/rclcpp/commit/5a88b09bfe0ad7f068f5d53b7f6de04f69f58a65)

For ease of reviewing, one thing I've done in the past is to have this PR target the branch of the PR it depends upon (that is, this PR would target brawner/allocator-memory-strategy-tests-p1).  Then when that PR gets merged, you rebase this one on top of it.

For now I'm just going to wait on reviewing this one, since some of the changes that come in via #1197 may need to be reflected here as well.

**Comment by wjwwood** [comment]
@clalancette is right, but I reviewed the actual code changes and they looked ok.

**Comment by brawner** [comment]
Rebased on to #1197 after updating over there and retargeting this PR to that branch.

I've had other organizations do it differently. I personally prefer it your way, but got in the habit of doing it how this was opened initially. That being said, Gerrit handles this the best that I've seen so far.

**Comment by brawner** [comment]
After merging #1227, retargeted and rebased on to master.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11450/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6665/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9392/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11352/)

File: rclcpp/test/rclcpp/strategies/test_allocator_memory_strategy.cpp [hunk 1]
```diff
@@ -351,6 +351,70 @@ class TestAllocatorMemoryStrategy : public ::testing::Test
...
+    auto basic_node = std::make_shared<rclcpp::Node>("basic_node", "ns");
+    nodes.push_back(node_with_entity1->get_node_base_interface());
+    nodes.push_back(basic_node->get_node_base_interface());
+    allocator_memory_strategy()->collect_entities(nodes);
+
+    rclcpp::AnyExecutable result = get_next_entity_func(nodes);
+    if (result.node_base != node_with_entity1->get_node_base_interface()) {
+      return ::testing::AssertionFailure() <<
+             "Failed to get expected entity with specified get_next_*() function";
+    }
```
Comment (peterpena) [inline rclcpp/test/rclcpp/strategies/test_allocator_memory_strategy.cpp:369]: Hello @brawner , is there a particular reason this test checks for insertion ordering of service handles? The reason I am asking, in PR #1218 ,  a map is used to store callback groups and nodes, so the assumption of insertion order is invalid in this case.

Comment (brawner) [reply rclcpp/test/rclcpp/strategies/test_allocator_memory_strategy.cpp:369]: So this file was just updated in #1252, but this part wasn't touched. However, if the iteration order cannot be guaranteed, then `basic_node` here should be created with the new method `create_node_with_disabled_callback_groups`. That should cause it to be skipped if it's visited first.

Comment (peterpena) [reply rclcpp/test/rclcpp/strategies/test_allocator_memory_strategy.cpp:369]: Yes that worked. Thanks @brawner . I wanted to make sure that by removing the assumption of iteration order, I was not killing the test.

**Pull request merged**